### DISCUSSION TITLE
Does metals have Scala 3 formatting support?

### DISCUSSION THREAD

**Discussion by jinalskothari** [scalameta/metals#2744] [Q&A] [answered]
Hello,

It is not clear to me if the latest release of Metals is supposed to be able to format Scala 3 code.

For context, I am unfamiliar with Scala ecosystem and am just getting into it. Reasonably experienced with software dev in general.

My attempts to make scala 3 code formatting work on my machine have failed. When a file has Scala 2 compatible syntax, it gets formatted. But it fails for syntax that I believe is scala 3 only.

For example this content in a file 
```scala
@main def hello: Unit = {
    println(msg)
    println("Hello !")    
}
def msg = "I was compiled by Scala 3. :)"
```
produces this Metals log when format document is executed.
```
2021.04.27 19:24:20 ERROR scalafmt: /Volumes/JinalData/Projects/scala3genbymetals/src/main/scala/Main.scala:1: error: expected start of definition
@main def hello: Unit = {
      ^
```

With the following code in the same file instead, formatting does work.
```scala
object Main extends App {
  println("Hello, World!")

}
``` 

Thank you

**Comment by dos65** [comment, marked as answer]
@jinalskothari Currently, it doesn't work out of the box and we are going to fix it soon.

For now, to make it work you need to specify the latest RC version and `scala3` dialect in  `.scalafmt.conf` - see an [example](https://github.com/scalameta/metals-feature-requests/issues/122#issuecomment-825654542)

**Reply by jinalskothari** [reply]
Thank you, @dos65! 
 
Me and a colleague had been struggling with this since a few hours.

Would it be worthwhile to add this to the docs somewhere in the meantime to help out new people like us?

**Reply by tgodzik** [reply]
We plan to add some additional UX fixes that will make sure that proper dialect is filled in by default and also make sure that a warning is shown if formatting without the proper dialect. We might be able to provide an automatic fix here too.

**Reply by jinalskothari** [reply]
I see, thank you.
And thank you for your contributions. 🙏 

I am feeling so grateful that something so powerful like Scala and its ecosystem exists.
